nasa/ow_simulator
# Oceanwater 1006 - Material grid implemented as a Visual Plugin

## Linked Issues:
| EPIC ⚡ | [OCEANWATER-812](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-812) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-1006](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1006) |
| Github :octocat:  | # |

## Sibling PRs
[ow_europa - PR#21](https://github.com/nasa/ow_europa/pull/21)

## Context
This plugin does not do anything of use yet. This is just foundational work that I'd like to have eyes on. Eventually this visual plugin will be responsible for applying colors/albedo/texture to the terrain corresponding to materials that intersect the heightmap mesh. 

## Summary of Changes
* **MaterialDistributionPlugin** created in the new **ow_materials** package. 
* Created the foundation for volumetricly distributed materials
  * Axis-aligned grid that can be given corner boundaries in world coordinates and populates the volume in between with cells capable of containing a material ID. 
  * Material struct for grouping material parameters. 
  * Material database that is populated by a YAML file in **ow_materials** that defines each material's parameter. 

## Test
1. Switch both **ow_simulator** and **ow_europa** to this branch.
2. View a live feed of gzserver's log using `tail -F ~/.gazebo/server-####/default.log`
3. In another terminal launch the atacama_y1a world `roslaunch ow atacama_y1a.launch`
4. The following output should show in the log
```
(1660852776 254772976) MaterialDistributionPlugin: Material grid centered at (0.65, 0, 0.23) with dimensions 1.3 x 2 x 0.14 meters. 
(1660852776 260661465) MaterialDistributionPlugin: Materials database populated with 3 materials.
(1660852776 260695728) MaterialDistributionPlugin: Successfully loaded!
```
Feel free try other plugin and material parameters by modifying `ow_europa/models/atacama_y1a/model.sdf` or `ow_simulator/ow_materials/materials/default.yaml`, respectively.
